### PR TITLE
Upgrade pip to 20.3.3 and virtualenv to 20.4.0

### PR DESCRIPTION
**Pull request by amanda11**
Using pip 20.3.3 and legacy resolver

### CODE REVIEW

**Comment by armab** [comment]
You can check the https://hub.docker.com/repository/registry-1.docker.io/stackstorm/buildpack/builds/f12894d9-f84b-4842-a2c5-7fea9e8abd7e at Docker Hub.

Copy-paste error log:
```
Step 4/5 : RUN yum -y install python-devel && wget -qO - https://bootstrap.pypa.io/get-pip.py | python && pip install --upgrade "pip==20.3.3" && pip install setuptools virtualenv && rm -rf /root/.cache
---> Running in 55c806a1803a
Loaded plugins: fastestmirror, ovl
Determining fastest mirrors
* base: mirror.centos.org
* extras: mirror.centos.org
* updates: mirror.centos.org
Resolving Dependencies
--> Running transaction check
---> Package python-devel.x86_64 0:2.7.5-90.el7 will be installed
--> Processing Dependency: python(x86-64) = 2.7.5-90.el7 for package: python-devel-2.7.5-90.el7.x86_64
--> Processing Dependency: python2-rpm-macros > 3-30 for package: python-devel-2.7.5-90.el7.x86_64
--> Processing Dependency: python-rpm-macros > 3-30 for package: python-devel-2.7.5-90.el7.x86_64
--> Running transaction check
---> Package python.x86_64 0:2.7.5-89.el7 will be updated
---> Package python.x86_64 0:2.7.5-90.el7 will be an update
--> Processing Dependency: python-libs(x86-64) = 2.7.5-90.el7 for package: python-2.7.5-90.el7.x86_64
---> Package python-rpm-macros.noarch 0:3-34.el7 will be installed
---> Package python2-rpm-macros.noarch 0:3-34.el7 will be installed
--> Running transaction check
---> Package python-libs.x86_64 0:2.7.5-89.el7 will be updated
---> Package python-libs.x86_64 0:2.7.5-90.el7 will be an update
--> Finished Dependency Resolution
Dependencies Resolved
================================================================================
Package Arch Version Repository Size
================================================================================
Installing:
python-devel x86_64 2.7.5-90.el7 updates 399 k
Installing for dependencies:
python-rpm-macros noarch 3-34.el7 base 9.1 k
python2-rpm-macros noarch 3-34.el7 base 8.1 k
Updating for dependencies:
python x86_64 2.7.5-90.el7 updates 96 k
python-libs x86_64 2.7.5-90.el7 updates 5.6 M
Transaction Summary
================================================================================
Install 1 Package (+2 Dependent packages)
Upgrade ( 2 Dependent packages)
Total download size: 6.1 M
Downloading packages:
Delta RPMs disabled because /usr/bin/applydeltarpm not installed.
--------------------------------------------------------------------------------
Total 13 MB/s | 6.1 MB 00:00
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
Updating : python-libs-2.7.5-90.el7.x86_64 1/7
Updating : python-2.7.5-90.el7.x86_64 2/7
Installing : python-rpm-macros-3-34.el7.noarch 3/7
Installing : python2-rpm-macros-3-34.el7.noarch 4/7
Installing : python-devel-2.7.5-90.el7.x86_64 5/7
Cleanup : python-2.7.5-89.el7.x86_64 6/7
Cleanup : python-libs-2.7.5-89.el7.x86_64 7/7
Verifying : python2-rpm-macros-3-34.el7.noarch 1/7
Verifying : python-libs-2.7.5-90.el7.x86_64 2/7
Verifying : python-devel-2.7.5-90.el7.x86_64 3/7
Verifying : python-2.7.5-90.el7.x86_64 4/7
Verifying : python-rpm-macros-3-34.el7.noarch 5/7
Verifying : python-libs-2.7.5-89.el7.x86_64 6/7
Verifying : python-2.7.5-89.el7.x86_64 7/7
Installed:
python-devel.x86_64 0:2.7.5-90.el7
Dependency Installed:
python-rpm-macros.noarch 0:3-34.el7 python2-rpm-macros.noarch 0:3-34.el7
Dependency Updated:
python.x86_64 0:2.7.5-90.el7 python-libs.x86_64 0:2.7.5-90.el7
Complete!
[91mTraceback (most recent call last):
[0m
[91m File "<stdin>", line 24244, in <module>
[0m
[91m File "<stdin>", line 199, in main
[0m
[91m File "<stdin>", line 82, in bootstrap
[0m
[91m File "[0m
[91m/tmp/tmpbsXUAQ/pip.zip/pip/_internal/cli/main.py[0m
[91m", line [0m
[91m60[0m
[91m
[0m
[91m [0m
[91msys.stderr.write(f"ERROR: {exc}")
[0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m [0m
[91m ^
[0m
[91mSyntaxError: invalid syntax
[0m
Removing intermediate container 55c806a1803a
The command '/bin/sh -c yum -y install python-devel && wget -qO - https://bootstrap.pypa.io/get-pip.py | python && pip install --upgrade "pip==20.3.3" && pip install setuptools virtualenv && rm -rf /root/.cache' returned a non-zero code: 1
```

**Comment by amanda11** [comment]
Putting back to draft whilst we fix why still getting complaints about no-site-packages. Also may want to use pip 20.3.3 as it matches the version used by virtualenv.

File: buildpack/centos7/Dockerfile [hunk 1]
```diff
@@ -43,11 +43,12 @@ RUN yum -y install \
...
-    wget -qO - https://bootstrap.pypa.io/get-pip.py | python && \
-    pip install --upgrade "pip>=19.0.0,<20.0.0" && \
-    pip install setuptools virtualenv && \
-    rm -rf /root/.cache
+RUN yum -y install python-devel
+RUN yum install -y python3 python3-devel
+RUN wget -qO - https://bootstrap.pypa.io/get-pip.py | python3
+RUN pip install --upgrade "pip==20.3.4"
+RUN pip install setuptools virtualenv
+RUN rm -rf /root/.cache
```
Comment (armab) [inline buildpack/centos7/Dockerfile:51]: Please revert the changes with several commands grouped behind a single `RUN`.
This is within Docker best practices to minimize the amount of layers and reduce the produced Docker image size.

**Review by armab — CHANGES_REQUESTED** [review]

File: buildpack/centos7/Dockerfile (hunk 1 above)
Comment (blag) [reply buildpack/centos7/Dockerfile]: Done.

**Review by cognifloyd — CHANGES_REQUESTED** [review]
The buildpack images were removed in #101 (614b4610c89cd8b01f855321bd25cf85845b7fcd and c345407118a1c6d9dd8f5332163479b2dfab4263) but they were reintroduced as empty files in this merge commit: 973b39a12ad3e198d18f5467d35c980884b108b3 So, please delete:
- buildpack/*/Dockerfile
- packagingbuild/Dockerfile.debian
- packagingbuild/Dockerfile.template

**Comment by amanda11** [comment]
> The buildpack images were removed in #101 (614b4610c89cd8b01f855321bd25cf85845b7fcd and c345407118a1c6d9dd8f5332163479b2dfab4263) but they were reintroduced as empty files in this merge commit: 973b39a12ad3e198d18f5467d35c980884b108b3 So, please delete:
> 
> - buildpack/*/Dockerfile
> 
> - packagingbuild/Dockerfile.debian
> 
> - packagingbuild/Dockerfile.template

Thanks for picking that up - will fix

File: packagingbuild/bionic/Dockerfile [hunk 2]
```diff
@@ -47,17 +47,26 @@ RUN apt-get update && \
...
+#
+# We use our dh-virtualenv version, since it fixes shebangd lines rewrites.
+# The shebangd rewrites are in master branch, but that requires dependency
+# on debhelper 12.
+#
 RUN apt-get -y install \
         python-virtualenv python-setuptools python-mock && \
         apt-get clean && \
-        git clone --branch stackstorm_patched_use_custom_pip https://github.com/StackStorm/dh-virtualenv.git /tmp/dh-virtualenv && \
+        git clone --branch rm_site_packages_upgrade_pip https://github.com/StackStorm/dh-virtualenv.git /tmp/dh-virtualenv && \
```
Comment (Kami) [inline packagingbuild/bionic/Dockerfile:66]: Per my comment on Slack - please open PR with your changes against ``stackstorm_patched_`` branch in dh-virtualenv repo and let's just use ``stackstorm_patched`` branch.

File: packagingbuild/xenial/Dockerfile [hunk 3]
```diff
@@ -53,14 +53,23 @@ RUN apt-get update && \
...
 # Install fresh pip and co
-RUN curl https://bootstrap.pypa.io/get-pip.py | python3.6 - virtualenv==16.6.0 pip==20.0.2 wheel setuptools; \
+RUN curl https://bootstrap.pypa.io/get-pip.py | python3.6 - virtualenv==20.4.0 pip==20.3.3 wheel setuptools; \
+
       pip3.6 install --upgrade requests[security] && rm -rf /root/.cache
 
+# This line just busts Docker's cache so it re-runs the next line
+# The GitHub API will return different results when the branch HEAD changes
+# See https://stackoverflow.com/a/39278224
+# ADD https://api.github.com/repos/StackStorm/dh-virtualenv/git/refs/heads/rm_site_packages_upgrade_pip version.json
```
Comment (Kami) [inline packagingbuild/xenial/Dockerfile:63]: One option would also to be pin to check out a specific commit revision in the git clone / checkout line

**Review by Kami — APPROVED** [review]
I approved this PR, but let's please also make the dh-virtualenv change mentioned in the comment.

**Comment by cognifloyd** [comment]
> > The buildpack images were removed in #101 ([614b461](https://github.com/StackStorm/st2packaging-dockerfiles/commit/614b4610c89cd8b01f855321bd25cf85845b7fcd) and [c345407](https://github.com/StackStorm/st2packaging-dockerfiles/commit/c345407118a1c6d9dd8f5332163479b2dfab4263)) but they were reintroduced as empty files in this merge commit: [973b39a](https://github.com/StackStorm/st2packaging-dockerfiles/commit/973b39a12ad3e198d18f5467d35c980884b108b3) So, please delete:
> > 
> > * buildpack/*/Dockerfile
> > * packagingbuild/Dockerfile.debian
> > * packagingbuild/Dockerfile.template
> 
> Thanks for picking that up - will fix

You got rid of the buildpack empty files. :+1:
Could you also remove the empty packagingbuild files?

**Comment by amanda11** [comment]
> > > The buildpack images were removed in #101 ([614b461](https://github.com/StackStorm/st2packaging-dockerfiles/commit/614b4610c89cd8b01f855321bd25cf85845b7fcd) and [c345407](https://github.com/StackStorm/st2packaging-dockerfiles/commit/c345407118a1c6d9dd8f5332163479b2dfab4263)) but they were reintroduced as empty files in this merge commit: [973b39a](https://github.com/StackStorm/st2packaging-dockerfiles/commit/973b39a12ad3e198d18f5467d35c980884b108b3) So, please delete:
> > > 
> > > * buildpack/*/Dockerfile
> > > * packagingbuild/Dockerfile.debian
> > > * packagingbuild/Dockerfile.template
> > 
> > 
> > Thanks for picking that up - will fix
> 
> You got rid of the buildpack empty files. 
> Could you also remove the empty packagingbuild files?

Oops - missed them. Thanks - hopefully got them all now!

**Review by cognifloyd — APPROVED** [review]

**Comment by amanda11** [comment]
TODO: Update the PR to use stackstorm-patched branch after the dh-virtualenv PR has merged changes into stackstorm-patched branch. Will push up that change after other PR merged....

File: packagingbuild/bionic/Dockerfile (hunk 2 above)
Comment (Kami) [reply packagingbuild/bionic/Dockerfile:66]: I assume you will update this line here when the other PR is merged, right?

**Comment by amanda11** [comment]
@armab Would you be able to re-review, as I believe your review comments have now been addressed. Thanks.

File: packagingbuild/bionic/Dockerfile (hunk 2 above)
Comment (amanda11) [reply packagingbuild/bionic/Dockerfile]: @Kami Yes - updated now - to use stackstorm_patched.

**Review by Kami — APPROVED** [review]
LGTM 👍

File: packagingbuild/bionic/Dockerfile [hunk 4]
```diff
@@ -47,17 +47,26 @@ RUN apt-get update && \
...
-# We use our dh-virtualenv version, since it fixes shebangd lines rewrites
+# This line just busts Docker's cache so it re-runs the next line
+# The GitHub API will return different results when the branch HEAD changes
+# See https://stackoverflow.com/a/39278224
+# ADD https://api.github.com/repos/StackStorm/dh-virtualenv/git/refs/heads/stackstorm_patched version.json
+# RUN rm version.json
+#
+# We use our dh-virtualenv version, since it fixes shebangd lines rewrites.
+# The shebangd rewrites are in master branch, but that requires dependency
+# on debhelper 12.
```
Comment (armab) [inline packagingbuild/bionic/Dockerfile:61]: I think in the future it's OK to switch to debhelper `12`.

From what I see we're on `10` right now: https://github.com/StackStorm/st2-packages/blob/master/packages/st2/debian/compat

**Review by armab — APPROVED** [review]
Looks good :+1: 

Thanks, everyone!

File: packagingbuild/bionic/Dockerfile (hunk 4 above)
Comment (amanda11) [reply packagingbuild/bionic/Dockerfile:61]: When I tried on bionic in our images, the latest debhelper it could download was 11. I don't think 12 is in the bionic repos.

Comment (armab) [reply packagingbuild/bionic/Dockerfile:61]: Makes sense, good to know!
So probably `10` for Ubuntu Xenial then.
